opampserver/pkg/server/server.go

```diff
@@ -0,0 +1,176 @@
+			http.Error(w, "Content-Type header is not application/x-protobuf", http.StatusBadRequest)
+			return
+		}
+
+		bytes, err := io.ReadAll(req.Body)
+		if err != nil {
+			http.Error(w, "Failed to read request body", http.StatusBadRequest)
+			return
+		}
+		// assume the data is not compressed, which is not relevant and not supported in odigos
+
+		// Decode the message as a Protobuf message.
+		var opampRequest protobufs.AgentToServer
+		err = proto.Unmarshal(bytes, &opampRequest)
+		if err != nil {
+			logger.Error(err, "Cannot decode opamp message from HTTP Body")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		deviceId, err := getDeviceIdFromHeader(req)
+		if err != nil {
+			logger.Error(err, "Failed to get device id from header")
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		var opampResponse *protobufs.ServerToAgent
+		connectionInfo, exists := connectionCache.GetConnection(deviceId)
+		if !exists {
+			connectionInfo, opampResponse, err = handlers.OnNewConnection(ctx, deviceId, &opampRequest)
+			if err != nil {
+				logger.Error(err, "Failed to process new connection")
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			if connectionInfo != nil {
+				connectionCache.AddConnection(deviceId, connectionInfo)
+			}
+		} else {
+
+			if opampRequest.AgentDisconnect != nil {
+				handlers.OnConnectionClosed(ctx, connectionInfo)
+				connectionCache.RemoveConnection(deviceId)
+			}
+
+			opampResponse, err = handlers.OnAgentToServerMessage(ctx, &opampRequest, connectionInfo)
+
+			if err != nil {
+				logger.Error(err, "Failed to process opamp message")
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+		}
+
+		err = handlers.PersistInstrumentationDeviceStatus(ctx, &opampRequest, connectionInfo)
+		if err != nil {
+			logger.Error(err, "Failed to persist instrumentation device status")
+			// still return the opamp response
+		}
+
+		if opampResponse == nil {
+			logger.Error(err, "No response from opamp handler")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// keep record in memory of last message time, to detect stale connections
+		connectionCache.RecordMessageTime(deviceId)
+
+		opampResponse.InstanceUid = opampRequest.InstanceUid
+
+		// Marshal the response.
+		bytes, err = proto.Marshal(opampResponse)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// Send the response.
+		w.Header().Set("Content-Type", "application/x-protobuf")
+		_, err = w.Write(bytes)
+
+		if err != nil {
+			logger.Error(err, "Failed to write response")
+		}
+	})
+
+	go func() {
+		if err := http.ListenAndServe(listenEndpoint, nil); err != nil {
+			logger.Error(err, "Error starting opamp server")
+		}
+	}()
+
+	go func() {
+		ticker := time.NewTicker(HeartbeatInterval)
+		defer ticker.Stop() // Clean up when done
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Info("Shutting down live connections timeout monitor")
```

Comment:
Is there any cleanup we need to do here in exit?

Reply:
good catch, added `server.Shutdown`